### PR TITLE
Add GPU skinning in RGL

### PR DESCRIPTION
**Pull request by msz-rai**
PR changes the way how skinned meshes are handled. Before, baked mesh vertices were updated and now RGL performs animation based on the bone pose.

*Note: Only Linux RGL binaries were updated*

### CODE REVIEW

File: Assets/RGLUnityPlugin/Scripts/LidarSensor.cs [hunk 1]
```diff
@@ -129,7 +129,7 @@ public void Awake()
 
         public void Start()
         {
-            sceneManager = FindObjectOfType<SceneManager>();
+            sceneManager = SceneManager.Instance;
             if (sceneManager == null)
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/RGLUnityPlugin/Scripts/LidarSensor.cs:133]: Still `SceneManager` can not be present on the scene. `Instance` initialization is performed in `MonoBehaviour::Awake()` method.

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/BonesPoseCacheManager.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace RGLUnityPlugin
+{
+    /// <summary>
+    /// Since multiple skinned meshes can share the same skeleton, BonesPoseCacheManager improves the performance of the simulation
+    /// by caching the calculated pose of the skeletons for the given simulation step.
+    /// </summary>
+    public static class BonesPoseCacheManager
+    {
+        class BonesPose
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/BonesPoseCacheManager.cs:27]: It is fine. `BonesPose` shouldn't be accessible outside `BonesPoseCacheManager`.

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/BonesPoseCacheManager.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            public SkinnedMeshRenderer smr; // needed to retrieve the current skeleton's pose
+            public Matrix4x4[] pose; // buffer for the skeleton's pose in the Unity representation
+            public float[] rglPose; // buffer for the skeleton's pose in the RGL representation
+            public int usageCount; // the counter on how many meshes use this skeleton
+            public int lastUpdateFrame = -1; // simulation frame
+            public int lastFixedUpdateFrame = -1; // physics cycle in the simulation frame
+        }
+
+        // The bone root acts as the identifier of the skeleton.
+        private static Dictionary<Transform, BonesPose> boneRootToBonesPose = new Dictionary<Transform, BonesPose>();
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/BonesPoseCacheManager.cs:47]: I have tested it and it worked fine for me. I guess it is serialized as a `Transform` component reference.

File: Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/BonesPoseCacheManager.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            else
+            {
+                boneRootToBonesPose[smr.rootBone].usageCount++;
+            }
+        }
+
+        public static void UnregisterBonesPoseInstance(Transform rootBone)
+        {
+            var bonesPose = boneRootToBonesPose[rootBone];
+            if (bonesPose is null)
```
[the remark this answers is not in the document]
Comment (msz-rai) [reply Assets/RGLUnityPlugin/Scripts/LowLevelWrappers/BonesPoseCacheManager.cs:64]: Good catch!

File: Assets/RGLUnityPlugin/Scripts/LidarSensor.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply Assets/RGLUnityPlugin/Scripts/LidarSensor.cs:133]: Fair point, you are right!

**Review by PawelLiberadzki — APPROVED** [review]
LGTM!

**Pull request merged**